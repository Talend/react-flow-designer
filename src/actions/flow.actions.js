import {
	FLOWDESIGNER_FLOW_ADD_ELEMENTS,
	FLOWDESIGNER_FLOW_RESET,
	FLOWDESIGNER_FLOW_LOAD,
} from '../constants/flowdesigner.constants';

/**
 * Ask to sequentially add elements to the flow, each creation should be checked against store,
 * then applied via current reducers
 *
 * @params {array} listOfActionCreation
 */
export const addFlowElements = listOfActionCreation => ({
	type: FLOWDESIGNER_FLOW_ADD_ELEMENTS,
	listOfActionCreation,
});

/**
 * ask for flow reset, emptying, nodes, links, ports collections
 */
export const resetFlow = () => ({
	type: FLOWDESIGNER_FLOW_RESET,
});

/**
 * load a new flow
 */
export const loadFlow = listOfActionCreation => ({
	type: FLOWDESIGNER_FLOW_LOAD,
	listOfActionCreation,
});
