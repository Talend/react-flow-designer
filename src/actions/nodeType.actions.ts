import { FLOWDESIGNER_NODETYPE_SET } from '../constants/flowdesigner.constants';

/**
 * Ask to set a map for nodeTypes
 * @param {Object} nodeTypes
 */
export const setNodeTypes = (nodeTypes: object) => ({
	type: FLOWDESIGNER_NODETYPE_SET,
	nodeTypes,
});

export default setNodeTypes;
