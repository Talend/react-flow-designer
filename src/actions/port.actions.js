/* @flow */

import type { PortGraphicalAttributes, Port, PortData, PortAction } from '../flow-typed';

/**
 * return an action to create a new port
 * @param {string} nodeId - identifier of the node to wich the created connector should be attached
 * @param {string} portId
 * @param {string} portType
 * @param {Object} attributes
 */
export function addPort(nodeId: string, portId: string, { data, graphicalAttributes }: {data: PortData, graphicalAttributes: PortGraphicalAttributes}): PortAction {
	return {
		type: 'FLOWDESIGNER_PORT_ADD',
		nodeId,
		portId,
		data,
		graphicalAttributes,
	};
}

export const addPorts = (nodeId: string, ports: Array<Port>): PortAction => ({
	type: 'FLOWDESIGNER_PORT_ADDS',
	nodeId,
	ports,
});

/**
 * return an action to set port attributes
 * @param {string} portId
 * @param {Object} graphicalAttributes
 */
export const setPortGraphicalAttribute = (portId: string, graphicalAttributes: PortGraphicalAttributes): PortAction => ({
	type: 'FLOWDESIGNER_PORT_SET_GRAPHICAL_ATTRIBUTES',
	portId,
	graphicalAttributes,
});

/**
 * Ask to remove an attribute on target port
 * @param {string} portId
 * @param {string} graphicalAttributesKey - the key of the attribute to be removed
 */
export const removePortGraphicalAttribute = (portId: string, graphicalAttributesKey: string): PortAction => ({
	type: 'FLOWDESIGNER_PORT_REMOVE_GRAPHICAL_ATTRIBUTES',
	portId,
	graphicalAttributesKey,
});

/**
 * return an action to set port attributes
 * @param {string} portId
 * @param {Object} graphicalAttributes
 */
export const setPortdata = (portId: string, data: Object): PortAction => ({
	type: 'FLOWDESIGNER_PORT_SET_DATA',
	portId,
	data,
});

/**
 * Ask to remove an attribute on target port
 * @param {string} portId
 * @param {string} datasKey - the key of the attribute to be removed
 */
export const removePortData = (portId: string, dataKey: string): PortAction => ({
	type: 'FLOWDESIGNER_PORT_REMOVE_DATA',
	portId,
	dataKey,
});

/**
 * return an action to remove port and all attached links
 * @param {string} portId
 */
export const removePort = (portId: string): PortAction => ({
	type: 'FLOWDESIGNER_PORT_REMOVE',
	portId,
});
