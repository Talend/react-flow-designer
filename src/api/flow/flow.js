import flow from 'lodash/flow';
import curry from 'lodash/curry';
import Immutable from 'immutable';

import { throwInDev } from '../throwInDev';
import { Node } from './..';

/**
 * check if node exist in flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {bool} true if node exist
 */
export const isNodeExist = curry((state, nodeId) => state.hasIn(['nodes', nodeId]));

const setOut = curry((nodeId, state) => state.setIn(['out', nodeId], new Immutable.Map()));
const deleteOut = curry((nodeId, state) => state.deleteIn(['out', nodeId]));

const setIn = curry((nodeId, state) => state.setIn(['in', nodeId], new Immutable.Map()));
const deleteIn = curry((nodeId, state) => state.deleteIn(['in', nodeId]));

const setChildrens = curry((nodeId, state) =>
	state.setIn(['childrens', nodeId], new Immutable.Map()),
);
const deleteChildrens = curry((nodeId, state) => state.deleteIn(['childrens', nodeId]));

const setParents = curry((nodeId, state) => state.setIn(['parents', nodeId], new Immutable.Map()));
const deleteParents = curry((nodeId, state) => state.deleteIn(['parents', nodeId]));

/**
 * add a node to the flow
 * @param {FlowState} state
 * @param {NodeRecord} node
 * @return {FlowState}
 */
export const addNode = curry((state, node) => {
	if (Node.isNodeElseThrow(node) && !isNodeExist(state, Node.getId(node))) {
		const nodeId = Node.getId(node);
		return flow([setOut(nodeId), setIn(nodeId), setChildrens(nodeId), setParents(nodeId)])(
			state.setIn(['nodes', Node.getId(node)], node),
		);
	}
	throwInDev(`Node with id = ${Node.getId(node)}, already exist, can't create node.`);
	return state;
});

/**
 * if exist remove a node from the flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {FlowState}
 */
export const deleteNode = curry((state, nodeId) => {
	if (isNodeExist(nodeId)) {
		return flow([
			deleteOut(nodeId),
			deleteIn(nodeId),
			deleteChildrens(nodeId),
			deleteParents(nodeId),
		])(state.deleteIn(['nodes', nodeId]));
	}
	return state;
});

/**
 * update a node
 * @param {FlowState} state
 * @param {string} nodeId
 * @param {NodeRecord} node
 * @return {FlowState}
 */
export const updateNode = curry((state, nodeId, node) => {
	if (Node.isNodeElseThrow(node) && (isNodeExist(state, Node.getId(node)), isNodeExist(nodeId))) {
		if (nodeId === Node.getId(node)) {
			return addNode(deleteNode(state, Node.getId(node)), node);
		}
		return state.setIn(['nodes', nodeId], node);
	}
	return state;
});
