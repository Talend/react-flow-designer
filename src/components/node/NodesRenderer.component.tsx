import PropTypes from 'prop-types';
import React from 'react';
import invariant from 'invariant';
import { mapOf } from 'react-immutable-proptypes';
import { Map } from 'immutable';

import { NodeType } from '../../constants/flowdesigner.proptypes';
import { NodeRecordMap, NodeRecord, Id, Position } from '../../customTypings/index.d';

type Props = {
	nodes: NodeRecordMap;
	nodeTypeMap: Map<string, any>;
	startMoveNodeTo: (nodeId: Id, nodePosition: string) => void;
	moveNodeTo: (nodeId: Id, nodePosition: Position) => void;
	moveNodeToEnd: (nodeId: Id, nodePosition: Position) => void;
	snapToGrid: boolean;
};

class NodesRenderer extends React.Component<Props> {
	static propTypes = {
		nodes: mapOf(NodeType).isRequired,
		nodeTypeMap: PropTypes.object.isRequired,
		startMoveNodeTo: PropTypes.func.isRequired,
		moveNodeTo: PropTypes.func.isRequired,
		moveNodeToEnd: PropTypes.func.isRequired,
		snapToGrid: PropTypes.bool.isRequired,
	};

	constructor(props: Props) {
		super(props);
		this.renderNode = this.renderNode.bind(this);
	}

	renderNode(node: NodeRecord) {
		const type = node.getNodeType();
		const ConcreteComponent = this.props.nodeTypeMap.get(type).component;
		if (!ConcreteComponent) {
			invariant(
				false,
				`<NodesRenderer />  the defined node type in your graph model hasn't been mapped into
				the dataflow configuration, check NodeType documentation`,
			);
		}
		return (
			<ConcreteComponent
				node={node}
				startMoveNodeTo={this.props.startMoveNodeTo}
				moveNodeTo={this.props.moveNodeTo}
				moveNodeToEnd={this.props.moveNodeToEnd}
				key={node.id}
				snapToGrid={this.props.snapToGrid}
			/>
		);
	}

	render() {
		return <g>{this.props.nodes.valueSeq().map(this.renderNode)}</g>;
	}
}

export default NodesRenderer;
