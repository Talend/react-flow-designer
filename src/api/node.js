/**
 * This module is public and deal with Graph's object Nodes
 */
import curry from 'lodash/curry';
import flow from 'lodash/flow';
import indexOf from 'lodash/indexOf';
import isString from 'lodash/isString';
import upperFirst from 'lodash/upperFirst';

import throwInDev from './throwInDev';
import { NodeRecord } from '../constants/flowdesigner.model';
import { isPositionElseThrow } from './position';
import { isSizeElseThrow } from './size';
import { Data } from './data';

const positionSelector = ['graphicalAttributes', 'position'];
const sizeSelector = ['graphicalAttributes', 'nodeSize'];
const componentTypeSelector = ['graphicalAttributes', 'nodeType'];

const FORBIDEN_GRAPHICAL_ATTRIBUTES = ['position', 'nodeSize', 'nodeType'];

/**
 * Test if the first parameter is a NodeRecord instance
 * @param {NodeRecord} node
 * @param {bool} doThrow - throw if not a node
 * @returns {bool}
 * @throws
 */
function isNode(node) {
	if (node && node instanceof NodeRecord) {
		return true;
	}
	return false;
}

/**
 * Test if the first parameter is a NodeRecord, throw if not
 * @param {*} node
 * @returns {bool}
 * @throws
 */
export function isNodeElseThrow(node) {
	const test = isNode(node);
	if (!test) {
		throwInDev(
			`Should be a NodeRecord was given ${node &&
				node.toString()}, you should use Node module functions to create and transform Nodes`,
		);
	}
	return test;
}

/**
 * @param {NodeRecord} node
 * @returns {string}
 */
function getId(node) {
	if (isNodeElseThrow(node)) {
		return node.get('id');
	}
	return null;
}

/**
 * @param {string} id
 * @param {NodeRecord}
 * @returns {NodeRecord}
 */
const setId = curry((id, node) => {
	if (isString(id) && isNodeElseThrow(node)) {
		return node.set('id', id);
	}
	throwInDev(`nodeId should be a string was given ${id && id.toString()}}`);
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getPosition(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(positionSelector);
	}
	return null;
}

/**
 * @param {PositionRecord} position
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setPosition = curry((position, node) => {
	if (isPositionElseThrow(position) && isNodeElseThrow(node)) {
		return node.setIn(positionSelector, position);
	}
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getSize(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(sizeSelector);
	}
	return null;
}

/**
 * @param {SizeRecord} size
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setSize = curry((size, node) => {
	if (isSizeElseThrow(size) && isNodeElseThrow(node)) {
		return node.setIn(sizeSelector, size);
	}
	return node;
});

function getComponentType(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(componentTypeSelector);
	}
	return null;
}

const setComponentType = curry((nodeType, node) => {
	if (isString(nodeType) && isNodeElseThrow(node)) {
		return node.setIn(componentTypeSelector, nodeType);
	}
	throwInDev(`nodeType should be a string was given ${nodeType && nodeType.toString()}`);
	return node;
});

/**
 * @param {String} key
 * @param {any} value
 * @param {nodeRecord} node
 * @returns {nodeRecord}
 */
const setData = curry((key, value, node) => {
	if (isNodeElseThrow(node)) {
		return node.set('data', Data.set(key, value, node.get('data')));
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return Data.get(key, node.get('data'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return Data.has(key, node.get('data'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return node.set('data', Data.delete(key, node.get('data')));
	}
	return node;
});

/**
 * given a key check if that key is white listed
 * @param {String} key
 * @returns {Bool}
 */
function isWhiteListAttribute(key) {
	if (indexOf(FORBIDEN_GRAPHICAL_ATTRIBUTES, key)) {
		return true;
	}
	throwInDev(
		`${key} is a protected value of the Node, please use get${upperFirst(key)} set${upperFirst(
			key,
		)} from this module to make change on those values`,
	);
	return false;
}

/**
 * @param {String} key
 * @param {any} value
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setGraphicalAttribute = curry((key, value, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set(
			'graphicalAttributes',
			Data.set(key, value, node.get('graphicalAttributes')),
		);
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return Data.get(key, node.get('graphicalAttributes'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return Data.has(key, node.get('graphicalAttributes'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set('graphicalAttributes', Data.delete(key, node.get('graphicalAttributes')));
	}
	return node;
});

/**
 * Create a new Node
 * @param {String} id
 * @param {PositionRecord} position
 * @param {SizeRecord} size
 * @param {String} componentType
 * @returns {NodeRecord}
 */
const create = curry((id, position, size, componentType) =>
	flow([setId(id), setPosition(position), setSize(size), setComponentType(componentType)])(
		new NodeRecord(),
	),
);

export const Node = {
	create,
	isNode,
	getId,
	getPosition,
	setPosition,
	getSize,
	setSize,
	getComponentType,
	setComponentType,
	setData,
	getData,
	hasData,
	deleteData,
	setGraphicalAttribute,
	getGraphicalAttribute,
	hasGraphicalAttribute,
	deleteGraphicalAttribute,
};
