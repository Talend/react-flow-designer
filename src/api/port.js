import curry from 'lodash/curry';
import flow from 'lodash/flow';
import Immutable from 'immutable';
import { PortRecord, PositionRecord } from '../constants/flowdesigner.model';

const positionSelector = ['graphicalAttributes', 'position'];
const componentTypeSelector = ['graphicalAttributes', 'portType'];
const portTopologySelector = ['graphicalAttributes', 'properties', 'type'];
const indexSelector = ['graphicalAttributes', 'properties', 'index'];

export function isPositionRecord(position, doThrow) {
	if (position && position instanceof PositionRecord) {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a PositionRecord was given ${position.toString()}`);
	}
	return false;
}

/**
 * Test if the first parameter is a PortRecord
 * @param {*} port
 * @param {bool} doThrow - throw if not a port
 * @return {bool}
 * @throws
 */
export function isPortRecord(port, doThrow = false) {
	if (port && port instanceof PortRecord) {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a PortRecord was given ${port.toString()}`);
	}
	return false;
}

/**
 * Test if the first parameter is a PortRecord, throw if not
 * @param {*} port
 * @return {bool}
 * @throws
 */
export function isPortRecordElseThrow(port) {
	return isPortRecord(port, true);
}

export function isTypology(typology, doThrow) {
	if (typology === 'SOURCE' || typology === 'SINK') {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a typology 'SOURCE' or 'SINK' was given ${typology.toString()}`);
	}
	return false;
}

export function getId(port) {
	const whatver = isPortRecordElseThrow(port);
	if (whatver) {
		return port.get('id');
	}
	return false;
}

const setId = curry((id, port) => {
	if (isPortRecord(port) && typeof id === 'string') {
		return port.set('id', id);
	}
	throw new Error(`id should be a string was given ${id.toString()}`);
});

export function getNodeId(port) {
	if (isPortRecord(port, true)) {
		return port.get('nodeId');
	}
	return false;
}

export const setNodeId = curry((nodeId, port) => {
	if (isPortRecord(port, true) && typeof nodeId === 'string') {
		return port.set('nodeId', nodeId);
	}
	throw new Error(`nodeId should be a string was given ${nodeId.toString()}`);
});

export function getPosition(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(positionSelector);
	}
	return false;
}

export const setPosition = curry((position, port) => {
	if (isPortRecord(port, true) && isPositionRecord(position, true)) {
		return port.setIn(positionSelector, position);
	}
	return false;
});

export function getComponentType(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(componentTypeSelector);
	}
	return false;
}

export const setComponentType = curry((componentType, port) => {
	if (isPortRecord(port, true) && typeof componentType === 'string') {
		return port.setIn(componentTypeSelector, componentType);
	}
	throw new Error(`componentType should be a string was given ${componentType.toString()}`);
});

export function getTypology(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(portTopologySelector);
	}
	return false;
}
export const setTypology = curry((flowType, port) => {
	if (isPortRecord(port, true && isTypology(flowType))) {
		return port.setIn(portTopologySelector, flowType);
	}
	return false;
});
export function getIndex(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(indexSelector);
	}
	return false;
}
export const setIndex = curry((index, port) => {
	if (isPortRecord(port, true) && typeof index === 'number') {
		return port.setIn(indexSelector, index);
	}
	throw new Error(`index should be a number was given ${index.toString()}`);
});

export function getData(port) {
	if (isPortRecord(port)) {
		return port.get('data');
	}
	return false;
}

export const setData = curry((map, port) => {
	if (isPortRecord(port) && Immutable.Map.isMap(map)) {
		return port.set('data', map);
	}
	throw new Error(`data should be a Immutable.Map go ${map.toString()}`);
});

export const createPortRecord = curry((id, nodeId, index, typology) => {
	const create = flow([setId(id), setNodeId(nodeId), setIndex(index), setTypology(typology)]);
	return create(new PortRecord());
});
