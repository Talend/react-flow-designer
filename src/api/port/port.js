import curry from 'lodash/curry';
import flow from 'lodash/flow';
import indexOf from 'lodash/indexOf';
import isString from 'lodash/isString';
import isNumber from 'lodash/isNumber';
import upperFirst from 'lodash/upperFirst';

import { throwInDev, throwTypeError } from '../throwInDev';
import { PortRecord } from '../../constants/flowdesigner.model';
import { PORT_SOURCE, PORT_SINK } from '../../constants/flowdesigner.constants';
import { isPositionElseThrow } from '../position/position';
import * as Data from '../data/data';

const positionSelector = ['graphicalAttributes', 'position'];
const componentTypeSelector = ['graphicalAttributes', 'portType'];
const portTopologySelector = ['graphicalAttributes', 'properties', 'type'];
const indexSelector = ['graphicalAttributes', 'properties', 'index'];

/** in future properties should be removed from the react-flow-designer lib */
const FORBIDEN_GRAPHICAL_ATTRIBUTES = ['properties', 'portType'];

/**
 * @typedef {Immutable.Record} LinkRecord
 */

/**
 * Test if the first parameter is a PortRecord instance
 * @param {Portrecord} port
 * @returns {bool}
 * @throws
 */
export function isPort(port) {
	if (port && port instanceof PortRecord) {
		return true;
	}
	return false;
}

/**
 * Test if the first parameter is a PortRecord, throw if not
 * @param {*} port
 * @returns {bool}
 * @throws
 */
export function isPortElseThrow(port) {
	const test = isPort(port);
	if (!test) {
		throwTypeError('PortRecord', port, 'port', 'Port');
	}
	return test;
}

/**
 * Check if the typology is one of the two accepted value
 * @param {*} typology
 */
export function isTypologyElseThrow(typology) {
	if (typology === PORT_SOURCE || typology === PORT_SINK) {
		return true;
	}
	throwInDev(
		`Should be a typology 'SOURCE' or 'SINK' was given ${typology && typology.toString()}`,
	);
	return false;
}

/**
 * @param {PortRecord} port
 * @returns {string}
 */
export function getId(port) {
	if (isPortElseThrow(port)) {
		return port.get('id');
	}
	return false;
}

/**
 * @param {string} id
 * @param {PortRecord} port
 * @returns {PortRecord}
 */
export const setId = curry((id, port) => {
	if (isString(id) && isPortElseThrow(port)) {
		return port.set('id', id);
	}
	throwInDev(`id should be a string was given ${id && id.toString()}`);
	return port;
});

/**
 * @param {PortRecord} port
 * @returns {string}
 */
export function getNodeId(port) {
	if (isPortElseThrow(port)) {
		return port.get('nodeId');
	}
	return false;
}

/**
 * @param {string} nodeId
 * @param {PortRecord} port
 * @returns {PortRecord}
 */
export const setNodeId = curry((nodeId, port) => {
	if (isString(nodeId) && isPortElseThrow(port)) {
		return port.set('nodeId', nodeId);
	}
	throwInDev(`nodeId should be a string was given ${nodeId && nodeId.toString()}`);
	return port;
});

/**
 * @param {PortRecord} port
 * @returns {PositionRecord}
 */
export function getPosition(port) {
	if (isPortElseThrow(port)) {
		return port.getIn(positionSelector);
	}
	return false;
}

/**
 * @param {PositionRecord} position
 * @param {PortRecord} port
 * @returns {Port}
 */
export const setPosition = curry((position, port) => {
	if (isPortElseThrow(port) && isPositionElseThrow(position)) {
		return port.setIn(positionSelector, position);
	}
	return false;
});

/**
 * @param {PortRecord} port
 * @returns {string}
 */
export function getComponentType(port) {
	if (isPortElseThrow(port)) {
		return port.getIn(componentTypeSelector);
	}
	return false;
}

/**
 * @param {string} componentType
 * @param {PortRecord} port
 * @returns {PortRecord}
 */
export const setComponentType = curry((componentType, port) => {
	if (isPortElseThrow(port) && isString(componentType)) {
		return port.setIn(componentTypeSelector, componentType);
	}
	throwInDev(
		`componentType should be a string was given ${componentType && componentType.toString()}`,
	);
	return port;
});

/**
 * @param {PortRecord} port
 * @returns {String}
 */
export function getTypology(port) {
	if (isPortElseThrow(port)) {
		return port.getIn(portTopologySelector);
	}
	return false;
}

/**
 * @param {string} typology
 * @param {PortRecord} port
 * @returns {PortRecord}
 */
export const setTypology = curry((typology, port) => {
	if (isPortElseThrow(port) && isTypologyElseThrow(typology)) {
		return port.setIn(portTopologySelector, typology);
	}
	return false;
});

/**
 * Index is set per port type and per port,
 * so the renderer can order ports visually
 * @param {PortRecord} port
 * @returns {number}
 */
export function getIndex(port) {
	if (isPortElseThrow(port)) {
		return port.getIn(indexSelector);
	}
	return false;
}

/**
 * @param {number} index
 * @param {PortRecord} port
 * @returns {PortRecord}
 */
export const setIndex = curry((index, port) => {
	if (isNumber(index) && isPortElseThrow(port)) {
		return port.setIn(indexSelector, index);
	}
	throwInDev(`index should be a number was given ${index && index.toString()}`);
	return port;
});

/**
 * @param {String} key
 * @param {any} value
 * @param {nodeRecord} port
 * @returns {nodeRecord}
 */
export const setData = curry((key, value, port) => {
	if (isPortElseThrow(port)) {
		return port.set('data', Data.set(key, value, port.get('data')));
	}
	return port;
});

/**
 * @param {String} key
 * @param {NodeRecord} port
 * @returns {any | null}
 */
export const getData = curry((key, port) => {
	if (isPortElseThrow(port)) {
		return Data.get(key, port.get('data'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} port
 * @returns {Bool}
 */
export const hasData = curry((key, port) => {
	if (isPortElseThrow(port)) {
		return Data.has(key, port.get('data'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} port
 * @returns {NodeRecord}
 */
export const deleteData = curry((key, port) => {
	if (isPortElseThrow(port)) {
		return port.set('data', Data.deleteKey(key, port.get('data')));
	}
	return port;
});

/**
 * given a key check if that key is white listed
 * @param {String} key
 * @returns {Bool}
 */
export function isWhiteListAttribute(key) {
	if (indexOf(FORBIDEN_GRAPHICAL_ATTRIBUTES, key) === -1) {
		return true;
	}
	throwInDev(
		`${key} is a protected value of the Port, please use get${upperFirst(key)} set${upperFirst(
			key,
		)} from this module to make change on those values`,
	);
	return false;
}

/**
 * @param {String} key
 * @param {any} value
 * @param {NodeRecord} port
 * @returns {NodeRecord}
 */
export const setGraphicalAttribute = curry((key, value, port) => {
	if (isPortElseThrow(port) && isWhiteListAttribute(key)) {
		return port.set(
			'graphicalAttributes',
			Data.set(key, value, port.get('graphicalAttributes')),
		);
	}
	return port;
});

/**
 * @param {String} key
 * @param {NodeRecord} port
 * @returns {any | null}
 */
export const getGraphicalAttribute = curry((key, port) => {
	if (isPortElseThrow(port) && isWhiteListAttribute(key)) {
		return Data.get(key, port.get('graphicalAttributes'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} port
 * @returns {Bool}
 */
export const hasGraphicalAttribute = curry((key, port) => {
	if (isPortElseThrow(port) && isWhiteListAttribute(key)) {
		return Data.has(key, port.get('graphicalAttributes'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} port
 * @returns {NodeRecord}
 */
export const deleteGraphicalAttribute = curry((key, port) => {
	if (isPortElseThrow(port) && isWhiteListAttribute(key)) {
		return port.set('graphicalAttributes', Data.deleteKey(key, port.get('graphicalAttributes')));
	}
	return port;
});
/**
 * minimal port creation factory, additionnals information can be set trought
 * the above set* functions
 * @param {string} id
 * @param {string} nodeId
 * @param {number} index
 * @param {string} typology
 * @param {string} componentType
 * @returns {PortRecord}
 */
export const create = curry((id, nodeId, index, typology, componentType) =>
	flow([
		setId(id),
		setNodeId(nodeId),
		setIndex(index),
		setTypology(typology),
		setComponentType(componentType),
	])(new PortRecord()),
);
