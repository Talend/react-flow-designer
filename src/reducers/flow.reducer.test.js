import { Map, OrderedMap } from 'immutable';

import { flowReducer } from './flow.reducer';
import * as nodeActions from '../actions/node.actions';
import * as portActions from '../actions/port.actions';
import { NodeRecord, PortRecord, SizeRecord, PositionRecord } from '../constants/flowdesigner.model';

describe('FLOWDESIGNER_FLOW_ADD_ELEMENTS is batching elements creation', () => {
	it('should batch one element creation', () => {
		expect(flowReducer(new Map(), {
			type: 'FLOWDESIGNER.FLOW.ADD_ELEMENTS',
			listOfActionCreation: [
				nodeActions.addNode(
					'nodeId',
					{ x: 10, y: 10 },
					{ height: 10, width: 10 },
					undefined,
					{}
				),
			],
		})).toEqual(new Map()
			.set('nodes', new Map()
				.set('nodeId', new NodeRecord({
					id: 'nodeId',
					position: new PositionRecord({
						x: 10,
						y: 10,
					}),
					nodeSize: new SizeRecord({
						height: 10,
						width: 10,
					}),
					nodeType: undefined,
					attr: new Map(),
				}))
			));
	});

	it('should batch many elements creation', () => {
		expect(flowReducer(new Map(), {
			type: 'FLOWDESIGNER.FLOW.ADD_ELEMENTS',
			listOfActionCreation: [
				nodeActions.addNode(
					'nodeId',
					{ x: 10, y: 10 },
					{ height: 10, width: 10 },
					undefined,
					{}
				),
				nodeActions.addNode(
					'node2',
					{ x: 10, y: 10 },
					{ height: 10, width: 10 },
					undefined,
					{}
				),
				portActions.addPort(
					'nodeId',
					'portId',
					undefined,
					{},
				),
			],
		})).toEqual(new Map()
			.set('nodes', new Map()
				.set('nodeId', new NodeRecord({
					id: 'nodeId',
					position: new PositionRecord({
						x: 10,
						y: 10,
					}),
					nodeSize: new SizeRecord({
						height: 10,
						width: 10,
					}),
					nodeType: undefined,
					attr: new Map(),
				}))
				.set('node2', new NodeRecord({
					id: 'node2',
					position: new PositionRecord({
						x: 10,
						y: 10,
					}),
					nodeSize: new SizeRecord({
						height: 10,
						width: 10,
					}),
					nodeType: undefined,
					attr: new Map(),
				}))
			).set('ports', new Map()
				.set('portId', new PortRecord({
					id: 'portId',
					nodeId: 'nodeId',
					portType: undefined,
					position: undefined,
					attr: new Map(),
				})))
		);
	});

	it('should handle throwing sub reducer by returning old state', () => {
		expect(flowReducer(new Map(), {
			type: 'FLOWDESIGNER.FLOW.ADD_ELEMENTS',
			listOfActionCreation: [
				nodeActions.addNode(
					'nodeId',
					{ x: 10, y: 10 },
					{ height: 10, width: 10 },
					undefined,
					{}
				),
				nodeActions.addNode(
					'node2',
					{ x: 10, y: 10 },
					{ height: 10, width: 10 },
					undefined,
					{}
				),
				portActions.addPort(
					'node3',
					'portId',
					undefined,
					{},
				),
			],
		})).toEqual(new Map());
	});
});
