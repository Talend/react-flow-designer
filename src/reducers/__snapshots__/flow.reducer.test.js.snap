exports[`FLOWDESIGNER_FLOW_ADD_ELEMENTS should batch many elements creation 1`] = `
Object {
  "childrens": Object {},
  "in": Object {},
  "links": Object {},
  "nodeTypes": Object {},
  "nodes": Object {},
  "out": Object {},
  "parents": Object {},
  "ports": Object {},
  "transform": Object {
    "k": 1,
    "x": 0,
    "y": 0,
  },
  "transformToApply": undefined,
}
`;

exports[`FLOWDESIGNER_FLOW_ADD_ELEMENTS should batch one element creation 1`] = `
Object {
  "childrens": Object {
    "nodeId": Object {},
  },
  "in": Object {
    "nodeId": Object {},
  },
  "links": Object {},
  "nodeTypes": Object {},
  "nodes": Object {
    "nodeId": Object {
      "data": Object {},
      "description": undefined,
      "graphicalAttributes": Object {
        "nodeSize": Object {
          "height": 10,
          "width": 10,
        },
        "position": Object {
          "x": 10,
          "y": 10,
        },
        "properties": Object {},
        "type": undefined,
      },
      "id": "nodeId",
      "label": undefined,
      "type": undefined,
    },
  },
  "out": Object {
    "nodeId": Object {},
  },
  "parents": Object {
    "nodeId": Object {},
  },
  "ports": Object {},
  "transform": Object {
    "k": 1,
    "x": 0,
    "y": 0,
  },
  "transformToApply": undefined,
}
`;

exports[`FLOWDESIGNER_FLOW_ADD_ELEMENTS should handle throwing sub reducer by returning old state 1`] = `
Object {
  "childrens": Object {},
  "in": Object {},
  "links": Object {},
  "nodeTypes": Object {},
  "nodes": Object {},
  "out": Object {},
  "parents": Object {},
  "ports": Object {},
  "transform": Object {
    "k": 1,
    "x": 0,
    "y": 0,
  },
  "transformToApply": undefined,
}
`;

exports[`FLOWDESIGNER_FLOW_LOAD should reset old flow state and load news not touching flow config should load elements 1`] = `
Object {
  "childrens": Object {},
  "in": Object {},
  "links": Object {},
  "nodeTypes": Object {},
  "nodes": Object {},
  "out": Object {},
  "parents": Object {},
  "ports": Object {},
  "transform": Object {
    "k": 1,
    "x": 0,
    "y": 0,
  },
  "transformToApply": undefined,
}
`;

exports[`FLOWDESIGNER_PAN_TO set a calculated transformation into transformToApply  1`] = `
Object {
  "childrens": Object {},
  "in": Object {},
  "links": Object {},
  "nodeTypes": Object {},
  "nodes": Object {},
  "out": Object {},
  "parents": Object {},
  "ports": Object {},
  "transform": Object {
    "k": 1,
    "x": 0,
    "y": 0,
  },
  "transformToApply": Transform {
    "k": 1,
    "x": -400,
    "y": -400,
  },
}
`;
